passes/pmgen/peepopt_muldiv_c.pmg

```diff
@@ -0,0 +1,83 @@
+pattern muldiv_c
+//
+// Authored by Akash Levy and Alain Dargelas of Silimate, Inc. under ISC license.
+// Transforms mul->div into const->mul when b and c are divisible constants:
+// y = (a * b_const) / c_const   ===>   a * eval(b_const / c_const)
+//
+
+state <SigSpec> a b_const mul_y
+
+match mul
+	// Select multiplier
+	select mul->type == $mul
+endmatch
+
+code a b_const mul_y
+	// Get multiplier signals
+	a = port(mul, \A);
+	b_const = port(mul, \B);
+	mul_y = port(mul, \Y);
+
+	// Fanout of each multiplier Y bit should be 1 (no bit-split)
+	if (nusers(mul_y) != 2)
+		reject;
+
+	// A and B can be interchanged
+	branch;
+	std::swap(a, b_const);
+endcode
+
+match div
+	// Select div of form (a * b_const) / c_const
+	select div->type == $div
+
+	// Check that b_const and c_const is constant
+	filter b_const.is_fully_const()
+	filter port(div, \B).is_fully_const()
+	index <SigSpec> remove_bottom_padding(port(div, \A)) === mul_y
+endmatch
+
+code
+	// Get div signals
+	SigSpec div_a = port(div, \A);
+	SigSpec c_const = port(div, \B);
+	SigSpec div_y = port(div, \Y);
+
+	// Get offset of multiplier result chunk in divider
+	int offset = GetSize(div_a) - GetSize(mul_y);
+
+	// Get properties and values of b_const and c_const
+	int b_const_width = mul->getParam(ID::B_WIDTH).as_int();
+	bool b_const_signed = mul->getParam(ID::B_SIGNED).as_bool();
+	bool c_const_signed = div->getParam(ID::B_SIGNED).as_bool();
+	int b_const_int = b_const.as_int(b_const_signed);
+	int c_const_int = c_const.as_int(c_const_signed);
+	int b_const_int_shifted = b_const_int << offset;
+
+	// Check that there are only zeros before offset
+	if (offset < 0 || !div_a.extract(0, offset).is_fully_zero())
+		reject;
+
+	// Check that b is divisible by c
+	if (b_const_int_shifted % c_const_int != 0)
+		reject;
+
+	// Check that every single output bit of the multiplier is connected consecutively to the div operator (offset accounted for)
+	for (int i = 0; i < mul_y.size(); i++) {
+		if (sigmap(mul_y[i]) != sigmap(div_a[i + offset])) {
+			reject;
+		}
+	}
+
+	// Rewire to only keep multiplier
+	mul->setPort(\B, Const(b_const_int_shifted / c_const_int, b_const_width));
```

Comment:
Because `b_const_int_shifted` is shifted up, there's no guarantee it fits into `b_const_width`

Also this might need special care wrt signedness. `B_SIGNED` on the multiplier may be false but `b_const_int_shifted / c_const_int` can be negative due to the divider. I wouldn't be opposed to requiring matching signedness on the divider and multiplier to make the transformation a little easier to reason about

Reply:
@povik should I change to: mul->setPort(\B, Const(b_const_int_shifted / c_const_int, b_const_width + offset));

Reply:
That would work but I suggest constructing the Const with 32 bits (the constructor default) and then calling compress(/signedness/) to fit to the constant. The reason for this is: unless `wreduce` is sequenced after `peepopt` I am not sure any pass would shrink the operand before we bitblast the multiplier

Reply:
Implemented here: https://github.com/akashlevy/yosys/pull/14/commits/8b08f81d192b085d200ed59b4d352c0b26a10b80
and 
https://github.com/akashlevy/yosys/pull/14/commits/8687e5f4fb2cce36c61df1097d48e5f336277a02